### PR TITLE
Fixed issue #1624

### PR DESCRIPTION
**Pull request by pgirikishore**
Resolves: *(direct link to the issue)*https://github.com/audacity/audacity/issues/1624

*(short description of the changes and the motivation to make the changes)*
The below solution was discussed and agreed upon by @SteveDaulton.
For a simple solution to the problem of batch processing being interrupted:
1. If the target bit rate is > 32 kbps and project rate <= 48 kHz, use the project rate (as now).
2. If the target bit rate is> 32 kbps and the project rate is > 48 kHz, use **48 kHz**.
3. If the target bit rate <= 32 kbps and project rate >  24 kHz, use **24 kHz**.
4. If the target bit rate <= 32 kbps and project rate <= 24 kHz, use the project rate (as now).

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behaviour changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behaviour

### CODE REVIEW

**Comment by SteveDaulton** [comment]
I've not yet tried building this.
Is the user still prompted in cases 2 and 3 when manually exporting? I think they should be, as the "automatic" sample rate selection may not be optimal in all cases (but definitely much better than now for batch processing).

**Comment by pgirikishore** [comment]
I was not aware that the user should still be prompted to select the sample rate in-case of manual exporting. Currently, the user is not prompted. I will look into that and fix that as well.

**Comment by Paul-Licameli** [comment]
The change of behavior should depend on the batch mode flag stored in AudacityProject.  When it is false, do not change behavior.

File: src/export/ExportMP3.cpp [hunk 1]
```diff
@@ -1859,7 +1858,17 @@ ProgressResult ExportMP3::Export(AudacityProject *project,
    // Verify sample rate
    if (!make_iterator_range( sampRates ).contains( rate ) ||
       (rate < lowrate) || (rate > highrate)) {
-      rate = AskResample(bitrate, rate, lowrate, highrate);
+       if (project->mBatch) {
```
Comment (Paul-Licameli) [inline src/export/ExportMP3.cpp:1861]: Does this compile?  The member is mBatchMode.

That would correct the first version of your fix, for the interactive case, in which behavior should not change.

Comment (pgirikishore) [reply src/export/ExportMP3.cpp:1861]: Apologies. I think I might have accidentally changed it before pushing the code. I will fix it.

**Comment by pgirikishore** [comment]
Hey @Paul-Licameli, Please review this PR when you have the time.

**Comment by Paul-Licameli** [comment]
You merged audacity/master into your branch -- which we don't recommend.  To keep up to date, it's better to fetch master, rebase your branch onto it, and force-push (git push -f) the branch to your fork.

But in this case, the changes are so small, I would rather you just use interactive rebase (git rebase -i) to "squash" all into one small commit.

Use multiple commits if there is complicated transformation of the code that should be broke into reviewable steps (not so here); but don't leave multiple commits that are a record of your corrections of earlier attempts that were not correct.

File: src/export/ExportMP3.cpp [hunk 2]
```diff
@@ -94,6 +94,7 @@
 #include "../widgets/AudacityMessageBox.h"
 #include "../widgets/ProgressDialog.h"
 #include "wxFileNameWrapper.h"
+#include "../../libraries/lib-project/Project.h"
```
Comment (Paul-Licameli) [inline src/export/ExportMP3.cpp:97]: Don't write that, simply write
```
#include "Project.h"
```

File: src/export/ExportMP3.cpp [hunk 3]
```diff
@@ -1859,7 +1859,17 @@ ProgressResult ExportMP3::Export(AudacityProject *project,
    // Verify sample rate
    if (!make_iterator_range( sampRates ).contains( rate ) ||
       (rate < lowrate) || (rate > highrate)) {
-      rate = AskResample(bitrate, rate, lowrate, highrate);
+       if (project->mBatchMode) {
+           if (bitrate > 32) {
+               rate = 48000;
+           }
+           else if (bitrate <= 32) {
+               rate = 240000;
```
Comment (Paul-Licameli) [inline src/export/ExportMP3.cpp:1867]: You meant 24000?

Comment (Paul-Licameli) [reply src/export/ExportMP3.cpp:1867]: Is the intention to default to `highrate` appropriate for the bit rate?  Then say so in a comment.

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]
I'm done with the review.  There is an error and should be some comments.

File: src/export/ExportMP3.cpp (hunk 3 above)
Comment (SteveDaulton) [reply src/export/ExportMP3.cpp:1867]: Perhaps clear without resorting to comments like this?
```
   // Verify sample rate
   if (!make_iterator_range( sampRates ).contains( rate ) ||
      (rate < lowrate) || (rate > highrate)) {
         if (project->mBatchMode) {
            if (rate == 0) {  // Maybe possible from macro command?
               return ProgressResult::Cancelled;
            }
            else if (rate < lowrate) {
               rate = lowrate;
            }
            else {
               rate = highrate;
            }
         }
         else {
            rate = AskResample(bitrate, rate, lowrate, highrate);
            if (rate == 0) {
               return ProgressResult::Cancelled;
         }
      }
   }

```

Comment (SteveDaulton) [reply src/export/ExportMP3.cpp:1867]: Thinking about it a bit more, I think it's possible to set an invalid MP3 bitrate via a macro, so to cover all possibilities we should ensure that macros also handle any *illegal* bitrate and not just rate == 0.

Comment (SteveDaulton) [reply src/export/ExportMP3.cpp:1867]: Yes it is possible for a macro to set an invalid sample rate, so something like...
```
   // Verify sample rate
   if (!make_iterator_range( sampRates ).contains( rate ) ||
      (rate < lowrate) || (rate > highrate)) {
         if (project->mBatchMode) {
            // Force valid sample rate in macros.
            if (!make_iterator_range( sampRates ).contains( rate )) {
               auto const bestRateIt = std::lower_bound(sampRates.begin(),
                                                         sampRates.end(), rate);
               rate = (bestRateIt == sampRates.end()) ? highrate : *bestRateIt;
               // wxLogDebug("Best Rate: %d", rate);
            }
            if (rate < lowrate) {
               rate = lowrate;
            }
            else if (rate > highrate) {
               rate = highrate;
            }
         }
         else {
            rate = AskResample(bitrate, rate, lowrate, highrate);
            if (rate == 0) {
               return ProgressResult::Cancelled;
         }
      }
   }

```

**Comment by Paul-Licameli** [comment]
Hello @pgirikishore , can you please do as I requested above:  squash all change into a single commit, rebase onto recent master (always rebase:  do NOT merge master into your work), then force-push?

**Pull request merged**